plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,241 @@
+    private static final int INTEGER_INSTANCE_SIZE = ClassLayout.parseClass(Integer.class).instanceSize();
+
+    private final FileSystemExchangeStorage exchangeStorage;
+    private final URI outputDirectory;
+    private final int outputPartitionCount;
+    private final int numBuffers;
+    private final Optional<SecretKey> secretKey;
+    private final int writeBufferSize;
+
+    private final Map<Integer, ExchangeStorageWriter> writers = new ConcurrentHashMap<>();
+    private final Map<Integer, SliceOutput> pendingBuffers = new ConcurrentHashMap<>();
+    private final LinkedBlockingQueue<SliceOutput> freeBuffers = new LinkedBlockingQueue<>();
+    private final AtomicReference<Throwable> failure = new AtomicReference<>();
+    private volatile boolean finished;
+    @GuardedBy("this")
+    private CompletableFuture<Void> blockedFuture = new CompletableFuture<>();
+
+    public FileSystemExchangeSink(FileSystemExchangeStorage exchangeStorage, URI outputDirectory, int outputPartitionCount, Optional<SecretKey> secretKey)
+    {
+        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+        this.numBuffers = outputPartitionCount * 2; // double buffering to overlap computation and I/O
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        this.writeBufferSize = exchangeStorage.getWriteBufferSizeInBytes();
+
+        for (int i = 0; i < numBuffers; ++i) {
+            freeBuffers.add(Slices.allocate(writeBufferSize).getOutput());
+        }
+    }
+
+    // This return result is more like a best-effort hint. The implementation is not thread-safe,
+    // but doesn't affect correctness. It's possible to return NOT_BLOCKED when it's actually
+    // blocked in certain cases. The decision is consciously made to avoid global locking.
+    @Override
+    public CompletableFuture<?> isBlocked()
+    {
+        if (freeBuffers.isEmpty() && pendingBuffers.size() < outputPartitionCount) {
+            synchronized (this) {
+                if (blockedFuture.isDone()) {
+                    blockedFuture = new CompletableFuture<>();
+                }
+                return blockedFuture;
+            }
+        }
+        else {
+            return NOT_BLOCKED;
+        }
+    }
+
+    @Override
+    public void add(int partitionId, Slice data)
+    {
+        throwIfFailed();
+
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!finished, "already finished");
+        if (finished) {
+            return;
+        }
+
+        writers.computeIfAbsent(partitionId, this::createWriter);
+        synchronized (writers.get(partitionId)) {
```

Comment:
why do we need `synchronized` here. Looks like the class cannot be called from multiple threads anyway.
E.g. `add` vs `finish` would not behave correctly (it looks like two `flushIfNeeded` could be called in parallel for same partition)

Reply:
`ExchangeStorageWriter` is not thread safe. We want `FileSystemExchangeSink` to be thread-safe, synchronize on a partition basis provides fine-grained locking.

@arhimondr Is it possible to call `add` and `finish` at the same time? Added partition-level synchronization in `finish` to address your comment.

Reply:
No, it's not. I'm going document it on the interface.

Reply:
We need to make this class thread-safe with fine-grained locking.